File: src/namagomi/minecraft/api/github/GitTree.ts
import fetch from 'electron-fetch'
import {IGitTree} from './IGitTree'
import {GetGitTrees} from './GetGitTrees'

export class GitTree implements IGitTree {
    data: { path: string, type: 'blob' | 'tree', sha: string, url: string }
    children: GitTree[]

    constructor() {
        this.data = {path: '', type: 'tree', sha: '', url: ''}
        this.children = []
    }

    public async build(owner: string, repo: string, sha: string, path?: string, type?: 'blob' | 'tree', url?: string) {
        this.data.sha = sha
        const apiUrl = `https://api.github.com/repos/${owner}/${repo}/git/trees/${sha}`
        if (path == null || type == null || url == null) {
            await this.fetchTree(owner, repo, apiUrl)
        } else {
            this.data.path = path
            this.data.type = type
            this.data.url = url
            if (type === 'tree')
                await this.fetchTree(owner, repo, url)
        }
        return this
    }

    private async fetchTree(owner: string, repo: string, apiUrl: string) {
        await fetch(apiUrl).then(p => p.json()).then(async (json: GetGitTrees) => {
            await Promise.all(json.tree.map(async (item) => {
                const path = item.path
                const type = item.type
                const sha = item.sha
                const url = item.url
                this.children.push(await new GitTree().build(owner, repo, sha, path, type, url))
            }))
        })
    }

    public async getAllPaths(): Promise<string[]> {
        const result: string[] = []
        await this.getAllPathsRecursive(result, '')
        return result
    }

    private async getAllPathsRecursive(result: string[], pwd: string): Promise<void> {
        if (this.data.type === 'blob') {
            result.push(pwd)
        } else {
            result.push(pwd)
            this.children.map(async (child: GitTree) => {
                await child.getAllPathsRecursive(result, `${pwd}/${child.data.path}`)
            })
        }
    }

    public async getAllFilePaths(): Promise<string[]> {
        const result: string[] = []
        await this.getAllFilePathsRecursive(result, '')
        return result
    }

    private async getAllFilePathsRecursive(result: string[], pwd: string): Promise<void> {
        if (this.data.type === 'blob') {
            result.push(pwd)
        } else {
            this.children.map(async (child: GitTree) => {
                await child.getAllFilePathsRecursive(result, `${pwd}/${child.data.path}`)
            })
        }
    }

    public async getAllDirectoryPaths(): Promise<string[]> {
        const result: string[] = []
        await this.getAllDirectoryPathsRecursive(result, '')
        return result
    }

    private async getAllDirectoryPathsRecursive(result: string[], pwd: string): Promise<void> {
        if (this.data.type === 'tree') {
            result.push(pwd)
            this.children.map(async (child: GitTree) => {
                await child.getAllDirectoryPathsRecursive(result, `${pwd}/${child.data.path}`)
            })
        }
    }

    public getData(path: string) {
        const paths = path.split('/').filter(p => p !== '')
        return paths.reduce(
            (tree: GitTree, path: string) => {
                const found = tree.children.find(
                    (value: GitTree) =>
                        value.data.path === path)
                if (found == undefined) {
                    throw new Error(`${paths} not found. no such file or directory`)
                } else {
                    return found
                }
            }, this)
    }

    public exists(path: string) {
        try {
            this.getData(path)
            return true
        } catch (e) {
            return false
        }
    }
}

File: src/main/scala/com/github/namagomi/main/github/GitTree.scala
package com.github.namagomi.main.github

import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.client.RequestBuilding.Get
import akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport
import akka.http.scaladsl.model.{HttpEntity, Uri}
import akka.http.scaladsl.unmarshalling.Unmarshal
import com.github.namagomi.main.github.Protocol._
import com.github.namagomi.main.github.GitTree.getFileType

import scala.concurrent.duration.Duration
import scala.concurrent.{Await, ExecutionContextExecutor}

class GitTree(val data: TreeData, var children: Seq[GitTree]) extends SprayJsonSupport {
  def this() = {
    this(TreeData("", FileType.Tree, "", ""), Seq.apply())
  }

  implicit val system: ActorSystem = ActorSystem()
  implicit val executionContext: ExecutionContextExecutor = system.dispatcher

  def build(owner: String, repo: String, sha: String): GitTree = {
    this.data.sha = sha
    val apiUrl = s"https://api.github.com/repos/$owner/$repo/git/trees/$sha"
    this.fetchTree(owner, repo, apiUrl)
    this
  }

  def build(owner: String, repo: String, sha: String, path: String, _type: FileType, url: String): GitTree = {
    this.data.sha = sha
    this.data.path = path
    this.data._type = _type
    this.data.url = url
    if (_type == FileType.Tree)
      this.fetchTree(owner, repo, url)
    this
  }

  def fetchTree(owner: String, repo: String, apiUrl: String): Unit = {
    val request = Get(Uri(apiUrl), HttpEntity.apply("application/json"))
    val response = Await.result(Http().singleRequest(request), Duration.Inf)
    val body = Await.result(Unmarshal(response.entity).to[GitTreesResponse], Duration.Inf)
    body.tree.foreach(item =>
      this.children = this.children :+ new GitTree().build(owner, repo, item.sha, item.path, getFileType(item.`type`), item.url)
    )
  }

  def getAllPaths(pwd: String = ""): Seq[String] = {
    this.data._type match {
      case FileType.Blob =>
        Seq(pwd)
      case FileType.Tree =>
        this.children.foldLeft(Seq(pwd))((a: Seq[String], child: GitTree) =>
          a ++ child.getAllPaths(s"$pwd/${child.data.path}")
        )
    }
  }

  def getAllFilePaths(pwd: String = ""): Seq[String] = {
    this.data._type match {
      case FileType.Blob =>
        Seq(pwd)
      case FileType.Tree =>
        this.children.foldLeft(Seq.empty[String])((a: Seq[String], child: GitTree) =>
          a ++ child.getAllPaths(s"$pwd/${child.data.path}")
        )
    }
  }

  def getAllDirPaths(pwd: String = ""): Seq[String] = {
    this.data._type match {
      case FileType.Blob =>
        Seq.empty[String]
      case FileType.Tree =>
        this.children.foldLeft(Seq(pwd))((a: Seq[String], child: GitTree) =>
          a ++ child.getAllPaths(s"$pwd/${child.data.path}")
        )
    }
  }

  def getData(path: String): Option[GitTree] = {
    val paths = path.split('/').filter(_ != "")
    paths.foldLeft(Option.apply(this))((tree, path) => {
      tree match {
        case Some(value) =>
          value.children.find(_.data.path == path) match {
            case Some(a) => Some(a)
            case None => None
          }
        case None => None
      }
    })
  }

  def exists(path: String): Boolean = {
    getData(path) match {
      case Some(_) => true
      case None => false
    }
  }
}

object GitTree {
  def getFileType(_type: String): FileType = {
    _type match {
      case "blob" => FileType.Blob
      case "tree" => FileType.Tree
    }
  }
}